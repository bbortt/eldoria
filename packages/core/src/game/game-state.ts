import type { DiceRoll } from './dice-roll';
import type { GameGrid } from './game-grid';

import type { Character } from '../stats';

export interface InitGameState {
  team: Character[];
  username: string;
}

export interface GameState extends InitGameState {
  diceRoll: DiceRoll;
  grid: GameGrid;
  startingPlayer?: string;
}
