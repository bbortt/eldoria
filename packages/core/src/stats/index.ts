export * from './character';
export * from './npc';
export * from './race';
export * from './specialization';
