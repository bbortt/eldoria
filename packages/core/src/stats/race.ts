export class Race {
  constructor(
    public label: string,
    public baseStrength: number,
    public baseDexterity: number,
    public baseConstitution: number,
    public baseIntelligence: number,
    public baseWisdom: number,
    public baseCharisma: number,
  ) {}

  static HUMAN = new Race('Human', 10, 10, 10, 10, 10, 10);
  static DWARF = new Race('Dwarf', 12, 8, 13, 10, 11, 9);
  static ELF = new Race('Elf', 8, 12, 9, 12, 11, 11);
  static HALFLING = new Race('Halfling', 8, 13, 10, 10, 11, 12);
  static GIANT = new Race('Giant', 14, 6, 13, 7, 10, 8);
}
