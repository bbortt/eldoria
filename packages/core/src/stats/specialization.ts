export class Specialization {
  constructor(
    public label: string,
    public hpBonus: number,
    public acBonus: number,
    public strengthMod: number,
    public dexterityMod: number,
    public constitutionMod: number,
    public intelligenceMod: number,
    public wisdomMod: number,
    public charismaMod: number,
  ) {}

  static TANK = new Specialization('Tank', 50, 3, 2, 0, 3, -1, 1, 0);
  static WARRIOR = new Specialization('Warrior', 30, 1, 3, 1, 1, 0, 0, 0);
  static ASSASSIN = new Specialization('Assassin', 10, 1, 1, 4, 0, 1, 1, 0);
  static ARCHER = new Specialization('Archer', 20, 0, 1, 3, 0, 1, 2, 0);
  static MAGE = new Specialization('Mage', 0, 0, -1, 1, 0, 4, 2, 1);
  static HEALER = new Specialization('Healer', 20, 0, 0, 0, 1, 1, 4, 1);
  static BUFFER = new Specialization('Buffer', 10, 0, 0, 1, 1, 2, 2, 2);
}
