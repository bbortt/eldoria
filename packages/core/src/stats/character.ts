import { Race } from './race';
import { Specialization } from './specialization';

export const newCharacter = (name: string, race: Race, specialization: Specialization): Character => {
  const dexterity = race.baseDexterity + specialization.dexterityMod;
  const ac = 10 + Math.floor(dexterity / 2) - 5 + specialization.acBonus;

  const constitution = race.baseConstitution + specialization.constitutionMod;
  const maxHP = 100 + specialization.hpBonus + constitution * 5;

  return new Character(
    name,

    race.label,
    specialization.label,

    maxHP,
    maxHP,
    ac,

    race.baseStrength + specialization.strengthMod,
    dexterity,
    constitution,
    race.baseIntelligence + specialization.intelligenceMod,
    race.baseWisdom + specialization.wisdomMod,
    race.baseCharisma + specialization.charismaMod,
  );
};

export class Character {
  constructor(
    public name: string,

    public race: string,
    public specialization: string,

    public currentHP: number,
    public maxHP: number,
    public ac: number,

    public strength: number,
    public dexterity: number,
    public constitution: number,
    public intelligence: number,
    public wisdom: number,
    public charisma: number,
  ) {}
}
