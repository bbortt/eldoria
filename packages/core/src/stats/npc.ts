import { Character } from './character';
import { Race } from './race';
import { Specialization } from './specialization';

const THANE = new Character('Thane', Race.DWARF, Specialization.TANK);
const NYSSA = new Character('Nyssa', Race.HUMAN, Specialization.ASSASSIN);
const ELYNDOR = new Character('Elyndor', Race.ELF, Specialization.MAGE);
const BROM = new Character('Brom', Race.HUMAN, Specialization.WARRIOR);
const SELENE = new Character('Selene', Race.ELF, Specialization.HEALER);
