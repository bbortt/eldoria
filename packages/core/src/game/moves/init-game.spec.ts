import { Character } from '../../stats';
import { GameState, InitGameState } from '../game-state';

import type { GameGrid } from '../game-grid';
import initGame from './init-game';

const mockGameState: GameState = {
  diceRoll: { '0': 1, '1': 2 },
  grid: {} as GameGrid,
  startingPlayer: null,
  team: [],
  username: '',
};

const mockGameConfiguration: InitGameState = {
  username: 'bbortt',
  team: [{} as Character],
};

describe('initGame', () => {
  it('should initialize the game state with the provided configuration', () => {
    const context = { G: mockGameState, ctx: { numPlayers: 2 } };

    // @ts-expect-error - type is not callable
    initGame(context, mockGameConfiguration);

    expect(context.G.username).toBe('bbortt');
    expect(context.G.team).toHaveLength(1);
    expect(Object.keys(context.G.diceRoll)).toMatchSnapshot(['0', '1']);
  });
});
