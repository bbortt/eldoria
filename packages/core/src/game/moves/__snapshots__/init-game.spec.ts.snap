// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initGame should initialize the game state with the provided configuration 1`] = `
[
  "0",
  "1",
]
`;
